build: Make config warnings fatal if -DWCONFIGURE_ERROR=ON

davidgumberg opened the pull request:
Summarizing #31476, the CentOS CI task has been configuring a build with python `3.9` (the latest version in the Stream 9 `BaseOS` repository) which is below the minimum version of `3.10` which resulted in a warning being appended to the `configure_warnings` cmake variable. https://github.com/bitcoin/bitcoin/blob/712cab3a8f8ad76db959337ddc35cb4c34cac388/CMakeLists.txt#L546-L553

This warning has been emitted by every CI run on CentOS[^1] since the minimum version was bumped (#30527) and until the centos task was bumped to stream 10 in #31593 But this has not resulted in CI failure, even though `utils` and `rpcauth` tests have been skipped.

This branch adds a cmake build option `WCONFIGURE_ERROR` that if on makes the presence of `configure_warnings` a `FATAL_ERROR` it also makes an incompatible BDB version a configure_warning if `-DWARN_INCOMPATIBLE_BDB=ON`.

[^1]: e.g. https://cirrus-ci.com/task/5526283183980544?logs=ci#L895, https://cirrus-ci.com/task/6597682371756032?logs=ci#L894

davidgumberg committed:
build: WCONFIGURE_ERROR makes config errors fatal.

Also log `configure_warnings` to cmake-configure-log before a
potentially fatal message to make CI output more readable.

hebasto commented:
Related to https://github.com/bitcoin/bitcoin/issues/31476.

fanquake commented:
Assuming the CI is always configured to use `-DWERROR`, this seems like the more logical/generic solution for #31476 compared to #31669. Although one thing that could be improved is making the error more clear, as in the current output, it's hidden far back in the scrollback.

fanquake commented:
> if we start using syntax features and behavior that are only available in versions greater than the minimum?

This should never be the case, otherwise the minimum required version isn't actually the minimum required version.

davidgumberg commented:
> Related to #31476.

🤦  I did not realize that's what #31593 was fixing.

> Assuming the CI is always configured to use `-DWERROR`, this seems like the more logical/generic solution for #31476 compared to #31669. Although one thing that could be improved is making the error more clear, as in the current output, it's hidden far back in the scrollback.

Done, by logging `configure_warnings` to the cmake-configure-log.

 > > if we start using syntax features and behavior that are only available in versions greater than the minimum?
> 
> This should never be the case, otherwise the minimum required version isn't actually the minimum required version.

Sorry, I misspoke, I meant syntax features and behavior that are only available in the minimum version or greater. Anyways, I think it's fine to leave it as a warning, since it doesn't impact the executables.

hebasto commented on CMakeLists.txt:
```diff
@@ -650,7 +655,8 @@ message("\n")
     message("  ******\n")
     foreach(warning IN LISTS configure_warnings)
-      message(WARNING "${warning}")
+      message(CONFIGURE_LOG "${warning}") # Log to the cmake-configure-log file before a potentially fatal message.
```
`message(CONFIGURE_LOG ...)` was [added](https://cmake.org/cmake/help/latest/command/message.html#configure-log) in CMake version 3.26. Our minimum required one is 3.22.

davidgumberg replied:
Thanks, fixed.

hebasto commented on CMakeLists.txt:
```diff
@@ -76,6 +76,14 @@ list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/module)
+set(configure_warnings)
+option(WERROR "Treat compiler warnings as errors." OFF)
+if(WERROR)
+    set(configure_warning_message_type FATAL_ERROR)
+else()
+    set(configure_warning_message_type WARNING)
```
style nit: 2 space indentation.

davidgumberg replied:
Thanks, fixed

hebasto reviewed (commented):
I don't like the idea of overloading the `WERROR` build option with additional functionality, as this is not what users would expect.

However, if this PR proceeds in its current direction, the new functionality must be clearly documented in the `WERROR` build option's help string.

> Although one thing that could be improved is making the error more clear, as in the current output, it's hidden far back in the scrollback.

There is no need to continue after encountering an error:
```diff
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -78,11 +78,13 @@ include(CMakeDependentOption)
 # When adding a new option, end the <help_text> with a full stop for consistency.
 set(configure_warnings)
 option(WERROR "Treat compiler warnings as errors." OFF)
-if(WERROR)
-    set(configure_warning_message_type FATAL_ERROR)
-else()
-    set(configure_warning_message_type WARNING)
-endif()
+macro(user_message message)
+  if(WERROR)
+    message(FATAL_ERROR ${message})
+  else()
+    list(APPEND configure_warnings ${message})
+  endif()
+endmacro()
 
 option(BUILD_DAEMON "Build bitcoind executable." ON)
 option(BUILD_GUI "Build bitcoin-qt executable." OFF)
@@ -116,7 +118,7 @@ if(WITH_BDB)
                     "BDB (legacy) wallets opened by this build will not be portable!"
     )
     if(WARN_INCOMPATIBLE_BDB)
-      list(APPEND configure_warnings
+      user_message(
         "Incompatible Berkeley DB found. If this is intended, pass \"-DWARN_INCOMPATIBLE_BDB=OFF\".\nPassing \"-DWITH_BDB=OFF\" will suppress this warning."
       )
     endif()
@@ -552,7 +554,7 @@ find_package(Python3 3.10 COMPONENTS Interpreter)
 if(Python3_EXECUTABLE)
   set(PYTHON_COMMAND ${Python3_EXECUTABLE})
 else()
-  list(APPEND configure_warnings
+  user_message(
     "Minimum required Python not found. Utils and rpcauth tests are disabled."
   )
 endif()
@@ -655,10 +657,7 @@ message("\n")
 if(configure_warnings)
     message("  ******\n")
     foreach(warning IN LISTS configure_warnings)
-      if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.26)
-        message(CONFIGURE_LOG "${warning}") # Log to the cmake-configure-log file before a potentially fatal message.
-      endif()
-      message(${configure_warning_message_type} "${warning}")
+      message(WARNING "${warning}")
     endforeach()
     message("  ******\n")
 endif()
```

fanquake commented:
> I don't like the idea of overloading the WERROR build option with additional functionality, as this is not what users would expect.

Can you elaborate on why this would be unexpected? This is an option to turn warnings into errors, and it'd just be turning more warnings into errors. I'm assuming any user who want to use this locally is exactly the kind of user who'd rather see more warnings than less?

hebasto commented:
> > I don't like the idea of overloading the WERROR build option with additional functionality, as this is not what users would expect.
> 
> Can you elaborate on why this would be unexpected? This is an option to turn warnings into errors, and it'd just be turning more warnings into errors.

Other projects use the same or similar build option only to turn _compiler's_ warnings into errors.

fanquake commented:
> Other projects use the same or similar build option only to turn compiler's warnings into errors.

Sure, but should that dictate what we do? Having an option to turn warnings into errors, that turns slightly more warnings into errors compared to other projects, and is mostly used in our CI where we want to surface all warnings, seems reasonable.  Given that CMake lacks any native functionality to acheive the same thing, using `-DWERROR` seems ok, especially if the alternative is implement `N` more build options (now/in future) with varying defaults for everything that may warn and needs to be caught.

luke-jr commented:
>should failing to meet the minimum python version just be a warning if we start using syntax features and behavior that are only available in versions equal to or greater than the minimum?

The minimum should probably not be bumped at all, until we are ready to start using features requiring it... and that usage is *why* we're disabling tests if the minimum isn't met, right?

hebasto commented:
As pointed in https://github.com/bitcoin/bitcoin/pull/31724#issuecomment-2610000303, this change will break builds on systems where a toolchain does not support PIE.

davidgumberg committed:
ci: -DWCONFIGURE_ERROR=ON in CI jobs

davidgumberg commented:
> Given that CMake lacks any native functionality to acheive the same thing, using `-DWERROR` seems ok, especially if the alternative is implement `N` more build options (now/in future) with varying defaults for everything that may warn and needs to be caught.

> I don't like the idea of overloading the `WERROR` build option with additional functionality, as this is not what users would expect.

> As pointed in [#31724 (comment)](https://github.com/bitcoin/bitcoin/pull/31724#issuecomment-2610000303), this change will break builds on systems where a toolchain does not support PIE.

In light of the fact that that reusing `WERROR` would make it impossible to build with `WERROR` on a platform that does not support PIE, and to avoid overloading `WERROR` with other unexpected behavior while maintaining a general solution, I've added 1 more build flag `WCONFIGURE_ERROR` which if on, makes the presence of `configure_warnings` an error, and enabled it in CI jobs.

---

> I don't think this is "worse". If the minimum version is 3.10, then any feature in that version should be allowed to be used.

> The minimum should probably not be bumped at all, until we are ready to start using features requiring it... and that usage is _why_ we're disabling tests if the minimum isn't met, right?

What I was trying to get at is:

Why maintain a list of tests that get disabled if the python version is below the minimum, if this list has not been maintained as newer python features have been used elsewhere. (e.g. match statement in `combine_logs.py`) and if the actual version of python required to run these tests is not even the minimum? (python 3.9, the minimum when the check was introduced)

It seems that either trying to run the functional tests with `ctest` should fail if the python version is below the minimum, or it should just proceed at the peril of the user, but I don't understand the present state where these tests are silently disabled.

maybe this discussion is more relevant in #31669

davidgumberg commented:
> > Summarizing #31476, the CentOS CI task has been configuring a build with python `3.9` (the latest version in the Stream 9 core repo) which is below the minimum version of `3.10`
> 
> I don't think this is entirely right. appstream ships with 3.11 and 3.12 on centos. (You can install them just like you can install g++-14 on centos 9).

By 'core repo' I meant what I guess CentOS calls the 'BaseOS' repository, fixed in the description.

> > Making this matter a bit worse, we now also [use a match statement](https://github.com/bitcoin/bitcoin/commit/c9fb38a590e3961e68e942d71f202e357466d15f)
> 
> I don't think this is "worse". If the minimum version is 3.10, then any feature in that version should be allowed to be used.

I was not trying to say that it is a bad thing that there are features from the minimum version being used, I was trying to motivate failure to meet minimum python requirements deserving stricter enforcement than a warning, but since this and the other question I asked about making a Python version below the minimum an error have been poorly communicated by me and slightly off-topic, I've removed them from the PR description.

davidgumberg commented:
Rebased to fix merge conflict.

s373nZ commented:
ACK https://github.com/bitcoin/bitcoin/pull/31665/commits/eaf6a5198d2b9033a6d6f0a4c01faddc67152857

Tested generation under the following variety of option combinations on Ubuntu 24.04.2 LTS:

1. `libdb++dev` package not installed: `cmake -B build -DWCONFIGURE_ERROR=ON -DWARN_INCOMPATIBLE_BDB=ON -DENABLE_WALLET=ON -DWITH_BDB=ON --fresh`
<details>
<summary>Test 1 result</summary>

```bash
CMake Error at /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find BerkeleyDB (missing: BerkeleyDB_LIBRARY
  BerkeleyDB_INCLUDE_DIR) (Required is at least version "4.8")
Call Stack (most recent call first):
  /usr/share/cmake-3.28/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/module/FindBerkeleyDB.cmake:107 (find_package_handle_standard_args)
  CMakeLists.txt:120 (find_package)
```
</details>

2. `libdb++dev` package installed (`5.3`): `cmake -B build -DWCONFIGURE_ERROR=ON -DWARN_INCOMPATIBLE_BDB=ON -DENABLE_WALLET=ON -DWITH_BDB=ON --fresh`
<details>
<summary>Test 2 result</summary>

```bash
-- Found BerkeleyDB: /usr/lib/x86_64-linux-gnu/libdb_cxx-5.3.so (found suitable version "5.3.28", minimum required is "4.8")
CMake Warning at CMakeLists.txt:123 (message):
  Found Berkeley DB (BDB) other than 4.8.

  BDB (legacy) wallets opened by this build will not be portable!
```
followed by
```
CMake Error at CMakeLists.txt:689 (message):
  Incompatible Berkeley DB found.  If this is intended, pass
  "-DWARN_INCOMPATIBLE_BDB=OFF".

  Passing "-DWITH_BDB=OFF" will suppress this warning.


-- Configuring incomplete, errors occurred!
```
</details>

3. `libdb++dev` package installed (`5.3`): `cmake -B build -DWCONFIGURE_ERROR=ON -DWARN_INCOMPATIBLE_BDB=OFF -DENABLE_WALLET=ON -DWITH_BDB=ON --fresh`
<details>
<summary>Test 3 result</summary>

```bash
CMake Warning at CMakeLists.txt:123 (message):
  Found Berkeley DB (BDB) other than 4.8.

  BDB (legacy) wallets opened by this build will not be portable!
```
followed by
```
-- Configuring done (11.4s)
-- Generating done (0.1s)
```
</details>

4. `libdb++dev` package not installed, use depends toolchain: `cmake -B build --toolchain depends/x86_64-pc-linux-gnu/toolchain.cmake -DWCONFIGURE_ERROR=ON -DWARN_INCOMPATIBLE_BDB=ON -DENABLE_WALLET=ON -DWITH_BDB=ON --fresh`

No warnings, no errors.

Although the goal with this flag is toward stricter CI, these tests explore behavior under some general conditions.

davidgumberg committed:
guix: -DWCONFIGURE_ERROR=ON for guix builds

purpleKarrot commented:
Concept NACK.

Please don't add nonstandard cmake variables (here: `WCONFIGURE_ERROR=ON`) for things that have builtin support in CMake. Just issue warnings with the [`message`](https://cmake.org/cmake/help/latest/command/message.html) command:

```cmake
message(WARNING "There is some potential issue on the system, like a new version of a dependency that has not been tested yet.")
message(AUTHOR_WARNING "There is a potention issue in cmake code, like suspicious arguments to a custom function")
message(DEPRECATION "You may be calling a custom function that is planned for removal, please migrate")
```

Clients can then control via command line flags like [`-Werror=<what>`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-Werror) what warnings they want to see hand whether they should be interpreted as error.

davidgumberg commented:
> Clients can then control via command line flags like [`-Werror=<what>`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-Werror) what warnings they want to see hand whether they should be interpreted as error.

Thanks for the suggestion, it seems like cmake's `-Werror=<what>` does not support making `WARNING` messages errors, only `AUTHOR_WARNING` with `-Werror=dev` and `DEPRECATION` with `-Werror=deprecated`.

For example:

```cmake
cmake_minimum_required(VERSION 3.10)
project(min)

message(WARNING "Warn")
```
Running cmake configuration with this example and `-Werror=dev` set succeeds without error:

```console
$ cmake -B build -Werror=dev
CMake Warning at CMakeLists.txt:4 (message):
  Warn


-- Configuring done (0.0s)
-- Generating done (0.0s)
-- Build files have been written to: /tmp/warningexample/build
```

davidgumberg commented:
Rebased to address reviewer feedback that the guix builds should have this option enabled, and that `CMakeLists.txt` uses 2 spaces for indentation, not 4.

davidgumberg commented:
Rebased to fix merge conflict, this needs conceptual review, it seems to me the main question is: Should it be possible to make config warnings fatal? Or is it the case that any time you want to make a warning fatal, you have a warning that should probably always be an error?

fanquake commented:
@davidgumberg FYI I've implemented something using `AUTHOR_WARNING`, as discussed above, in #32865.

davidgumberg commented on .github/ci-test-each-commit-exec.sh:
```diff
@@ -0,0 +1,21 @@
+echo "Running test-one-commit on $( git log -1 )"
+
+# Use clang++, because it is a bit faster and uses less memory than g++
+CC=clang CXX=clang++ cmake -B build -DWERROR=ON -DWCONFIGURE_ERROR=ON -DWITH_ZMQ=ON -DBUILD_GUI=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DWITH_USDT=ON -DCMAKE_CXX_FLAGS='-Wno-error=unused-member-function'
```
Thanks for catching this, I've fixed this for the sake of posterity, but I've closed this PR in favor of #32865.

davidgumberg replied:
OK, looks like the PR won't update with my branch now that it's closed, but it's here: https://github.com/davidgumberg/bitcoin/tree/1-15-24-configure_warnings
